### DISCUSSION TITLE
Closed loop mill spindle rpm on RT1062?

### DISCUSSION THREAD

**Discussion by mlewus** [terjeio/grblHAL#298] [Ideas]
I’m using the RT1062 driver for a 4 axis cnc mill. The spindle uses a 90VDC motor controlled via a 0-10vdc control signal derived (averaged) from GRBL’s PWM output. Rather than setting a PWM value for spindle speed I would like to configure GRBL with encoder pulses/rev, set a speed in RPM, and let it PID spindle PWM to get the correct speed. The controller can’t react faster than half wave US line frequency or 120Hz, so a PID update of 8ms or better would be ideal.

This is for general milling operations so it does not have to be nearly as accurate as spindle sync for threading. +/-10% of set point would be terrific.

Does this functionality exist, or is it planned?

**Comment by terjeio** [comment]
It kind of exists as I have experimented with this for the [MSP432P401 driver](https://github.com/grblHAL/MSP432P401R/blob/master/driver.c). Currently only this, iMXRT1062 (Teensy 4) and STM32F4xx drivers has support for the encoder input(s) and the low-level driver code required.

Adding the closed loop code (by copy/paste from the MSP432 driver) to iMXRT1062 and STM32F4xx should be fairly easy, other drivers not necessarily so as at least one free GPIO input for the encoder is required.

**Comment by mlewus** [comment]
I am building the mill controller now. When I get it working I'll take a look at the MSP432P401 driver and see if my creaky coding skills are up to the task of porting to Teensy 4.

Thanks for your quick response.

**Comment by mlewus** [comment]
@terjeio, I started reviewing the code and I have a few questions. In MSP432P401R/driver.h there are two defines, RPM_COUNTER_PIN (line 214) and RPM_INDEX_PIN (line 223). 
1. Are these the pin defs for the spindle encoder and the 0 degree index pulse, respectively?
2. This would imply a single phase encoder signal, not quadrature, correct?
3. Are the pulses active high or low?

**Reply by terjeio** [reply]
1. Yes - the index input is only required for spindle sync
2. Correct, quadrature input is only required for rigid tapping or synced motion when hand turning the spindle. Not going to happen anytime soon.
3. Does not matter much, the inputs are edge triggered.

The iMXRT1062 driver already has code for handling the encoder input so this part is done and dusted. Note that the core has two versions of providing spindle speed to the driver, either by a PWM value or a RPM value. For spindle sync/closed loop spindle the RPM variant is required. Enable by removing the comment from `#define SPINDLE_RPM_CONTROLLED` in _config/grbl.h_.

The PID control loop requires the sampling rate (for the actual RPM value) to be constant, since all grblHAL drivers has a 1ms timer interrupt going (from the systick timer in ARM cores) it is natural to hook the control loop calculations into its handler.

See [this discussion](https://github.com/terjeio/grblHAL/discussions/289) for info about modifying Phils breakout board, this since the only input useable for the encoder input is pin 14 unless the related code is rewritten. A modified board map file, `BOARD_T41U5XBB_SS`, is available at the [new driver repo](https://github.com/grblHAL/iMXRT1062).

**Reply by phil-barrett** [reply]
Mark has a Pro beta so he doesn't need to mod the board.

**Comment by mlewus** [comment]
@terjeio you previously mentioned that the encoder logic is edge triggered. For the purposes of determining pulses per revolution, does the encoder trigger on one edge or both edges?

**Reply by terjeio** [reply]
> For the purposes of determining pulses per revolution, does the encoder trigger on one edge or both edges?

On one edge only.
